datalad/tests/utils_cached_dataset.py

```diff
@@ -0,0 +1,233 @@
+    # have it, there's no need for an update. Otherwise it gets tricky, because
+    # this is a cache, not a checkout a test would operate on. It needs to
+    # behave as if it was the thing at `url` from the point of view of the test
+    # using it (cloning/getting content from here). We would nee to update all
+    # references, not just fetch them!
+    #
+    # Can we even (cheaply) tell whether `version` is an absolute reference
+    # (actual SHA, not a branch/tag)?
+    #
+    # NOTE: - consider git-clone --mirror, but as w/ --bare: not an option for
+    #         datalad-clone yet.
+    #       - --reference[-if-able] might also be worth thinking about for
+    #         the clone @cached_dataset creates wrt clone in cacheq
+    #
+    # So, for now fetch, figure whether there actually was something to fetch
+    # and if so simply invalidate cache and re-clone/get. Don't overcomplicate
+    # things. It's about datasets used in the tests - they shouldn't change too
+    # frequently.
+    elif any('uptodate' not in c['operations']
+             for c in ds.repo.fetch('origin')):
+        rmtree(ds.path)
+        ds = Clone()(url, ds.pathobj)
+
+    if version:
+        # check whether version is available
+        assert ds.repo.commit_exists(version)
+    if keys and AnnexRepo.is_valid_repo(ds.path):
+        ds.repo.get(keys, key=True)
+
+    return ds
+
+
+@optional_args
+def cached_dataset(f, url=None, version=None, paths=None):
+    """Test decorator providing a clone of `url` from cache
+
+    If config datalad.tests.cache is not set, delivers a clone in a temporary
+    location of the original `url`. Otherwise that clone is in fact a clone of a
+    cached dataset (origin being the cache instead of `url`).
+    This allows to reduce time and network traffic when using a dataset in
+    different tests.
+
+    The clone will checkout `version` and get the content for `paths`.
+
+    Parameters
+    ----------
+    url: str
+        URL to the to be cloned dataset
+    version: str
+        committish to checkout in the clone
+    paths: str or list
+        annexed content to get
+
+    Returns
+    -------
+    Dataset
+        a clone of the dataset at `url` at a temporary location (cleaned up,
+        after decorated test is finished - see with_tempfile). If caching is
+        enabled, it's actually a clone of a clone, 'origin' being the clone in
+        cache rather than the original repo at `url`.
+    """
+    @better_wraps(f)
+    @with_tempfile
+    def newfunc(*arg, **kw):
+
+        if DATALAD_TESTS_CACHE:
+            # Note: We can't pass keys based on `paths` parameter to
+            # get_cached_dataset yet, since translation to keys depends on a
+            # worktree. We'll have the worktree of `version` only after cloning.
+            ds = get_cached_dataset(url, version=version)
+            clone_ds = Clone()(ds.pathobj, arg[-1])
```

Comment:
BTW -- to be more "transparent cache" instance, shouldn't then it 
- clone so that local cached "origin" is named differently than "origin" remote (e.g. `local-cached`)
- add "origin" remote which would be the original URL?
the only possible side-effect I see could be -- it would need to fetch that "origin" remote so might prevent testing "offline" but so far I think the entire approach is for "online" testing

Reply:
also probably that would undermine benefits for some tests use cases which might want to request data specifically from "origin".

Reply:
>  add "origin" remote which would be the original URL?
the only possible side-effect I see could be -- it would need to fetch that "origin" remote so might prevent testing "offline" but so far I think the entire approach is for "online" testing

Quite the opposite! With that approach a test would need to be aware of the cache in order to make any use of it - that's the exact opposite of transparent. With the implementation here, the test would always talk to "origin" - not knowing whether or not it's cached. That is the point! Also: I can't disable caching without side effect on the tests, if they would explicitly talk to a cache remote.